storage/core/src/main/java/com/hortonworks/registries/storage/transaction/ManagedTransaction.java

```diff
@@ -0,0 +1,272 @@
+     * @param <T3> type for parameter3 of consumer
+     * @throws Exception
+     */
+    public <T1, T2, T3> void executeConsumer(ManagedTransactionConsumer.Arg3<T1, T2, T3> fn,
+                                             T1 t1, T2 t2, T3 t3) throws Exception {
+        executeTransactionBlockInternal((_void) -> {
+            fn.apply(t1, t2, t3);
+            return null;
+        });
+    }
+
+    /**
+     * Execute consumer with managed transaction.
+     *
+     * @param fn consumer to execute with managed transaction
+     * @param <T1> type for parameter1 of consumer
+     * @param <T2> type for parameter2 of consumer
+     * @param <T3> type for parameter3 of consumer
+     * @param <T4> type for parameter4 of consumer
+     * @throws Exception
+     */
+    public <T1, T2, T3, T4> void executeConsumer(ManagedTransactionConsumer.Arg4<T1, T2, T3, T4> fn,
+                                                 T1 t1, T2 t2, T3 t3, T4 t4) throws Exception {
+        executeTransactionBlockInternal((_void) -> {
+            fn.apply(t1, t2, t3, t4);
+            return null;
+        });
+    }
+
+    /**
+     * Execute consumer with managed transaction.
+     *
+     * @param fn consumer to execute with managed transaction
+     * @param <T1> type for parameter1 of consumer
+     * @param <T2> type for parameter2 of consumer
+     * @param <T3> type for parameter3 of consumer
+     * @param <T4> type for parameter4 of consumer
+     * @param <T5> type for parameter5 of consumer
+     * @throws Exception
+     */
+    public <T1, T2, T3, T4, T5> void executeConsumer(ManagedTransactionConsumer.Arg5<T1, T2, T3, T4, T5> fn,
+                                                     T1 t1, T2 t2, T3 t3, T4 t4, T5 t5) throws Exception {
+        executeTransactionBlockInternal((_void) -> {
+            fn.apply(t1, t2, t3, t4, t5);
+            return null;
+        });
+    }
+
+    private <R> R executeTransactionBlockInternal(FunctionCapableOfThrowingException<Void, R> fn) throws Exception {
+        boolean committed = false;
+        try {
+            transactionManager.beginTransaction(transactionIsolation);
+            R r = fn.apply(null);
+            transactionManager.commitTransaction();
+            committed = true;
+            return r;
+        } catch (IgnoreTransactionRollbackException e) {
+            transactionManager.commitTransaction();
+            committed = true;
+            throw convertThrowableToException(e.getCause());
+        } finally {
+            if (!committed) {
+                transactionManager.rollbackTransaction();
+            }
+        }
+    }
+
+    private Exception convertThrowableToException(Throwable t) {
+        if (t instanceof Error) {
+            throw (Error) t;
+        } else if (t instanceof Exception) {
+            return (Exception) t;
+        } else {
+            return new RuntimeException(t);
+        }
+    }
+
+    private interface FunctionCapableOfThrowingException<T, R> {
```

Comment:
nit: this can be a `SupplierCapableOfThrowingException<R>` since the only usage does not take any argument.

Reply:
Nice catch. Addressed.